File: go/signing/sign.go
package signing

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"fmt"
	"strconv"
	"strings"
	"time"
)

// Sign computes a single Standard Webhooks "v1,<sig>" signature.
//
//   - secret: the signing secret (with or without the whsec_ prefix)
//   - id: the webhook delivery id
//   - timestamp: the delivery timestamp
//   - payload: the exact raw JSON payload string
func Sign(secret string, id string, timestamp time.Time, payload string) (string, error) {
	timestampSeconds := timestamp.Unix()
	signedContent := fmt.Sprintf("%s.%d.%s", id, timestampSeconds, payload)
	key, err := ParseSecret(secret)
	if err != nil {
		return "", err
	}
	mac := hmac.New(sha256.New, key)
	mac.Write([]byte(signedContent))
	digest := base64.StdEncoding.EncodeToString(mac.Sum(nil))
	return "v1," + digest, nil
}

// Signer computes Standard Webhooks delivery headers, optionally signing with
// multiple secrets for zero-downtime key rotation.
type Signer struct {
	secrets []string
}

// NewSigner builds a Signer over one or more secrets (space-joined in the
// webhook-signature header, one v1,<sig> entry per secret).
func NewSigner(secrets ...string) *Signer {
	return &Signer{secrets: secrets}
}

// Headers computes the webhook-id, webhook-timestamp, and webhook-signature headers.
//
//   - id: the webhook delivery id
//   - payload: the exact raw JSON payload string
//   - timestamp: the delivery timestamp
func (s *Signer) Headers(id string, payload string, timestamp time.Time) (map[string]string, error) {
	timestampSeconds := timestamp.Unix()
	sigs := make([]string, 0, len(s.secrets))
	for _, secret := range s.secrets {
		sig, err := Sign(secret, id, timestamp, payload)
		if err != nil {
			return nil, err
		}
		sigs = append(sigs, sig)
	}
	return map[string]string{
		"webhook-id":        id,
		"webhook-timestamp": strconv.FormatInt(timestampSeconds, 10),
		"webhook-signature": strings.Join(sigs, " "),
	}, nil
}

// CoreSigner adapts Signer to the shape of anyhook-core's WebhookSigner port
// (Sign(secrets []string, id, payload string, timestampMs int64) map[string]string).
// Package core is never imported here (by design, G1): Go's structural
// interface satisfaction means CoreSigner implements core.WebhookSigner
// purely by matching its method signature, the same "structural" trick the
// TS package uses to avoid a hard dependency edge on @anyhook/core.
//
// Sign is infallible by port contract (mirrors the TS synchronous, non-throwing
// port): a malformed secret (e.g. invalid base64) is a configuration error
// that should be caught at secret-creation/rotation time, not at delivery
// time, so this adapter degrades to an empty signature for that entry rather
// than panicking or returning an error.
type CoreSigner struct{}

// Sign implements the core.WebhookSigner port shape.
func (CoreSigner) Sign(secrets []string, id string, payload string, timestampMs int64) map[string]string {
	ts := time.UnixMilli(timestampMs)
	hdrs, err := NewSigner(secrets...).Headers(id, payload, ts)
	if err != nil {
		return map[string]string{
			"webhook-id":        id,
			"webhook-timestamp": strconv.FormatInt(ts.Unix(), 10),
			"webhook-signature": "",
		}
	}
	return hdrs
}

File: packages/signing/src/sign.ts
/** @fileoverview Standard Webhooks signature generation. @module @anyhook/signing */
import { createHmac } from 'node:crypto';
import { parseSecret } from './secret.js';

/**
 * Computes a single Standard Webhooks `v1,<sig>` signature.
 * @param secret - The signing secret (with or without `whsec_` prefix)
 * @param id - The webhook delivery id
 * @param timestamp - The delivery timestamp
 * @param payload - The exact raw JSON payload string
 * @returns The `v1,<base64 signature>` string
 */
export function sign(secret: string, id: string, timestamp: Date, payload: string): string {
  const timestampSeconds = Math.floor(timestamp.getTime() / 1000);
  const signedContent = `${id}.${timestampSeconds}.${payload}`;
  const key = parseSecret(secret);
  const digest = createHmac('sha256', key).update(signedContent).digest('base64');
  return `v1,${digest}`;
}

/**
 * Computes Standard Webhooks delivery headers, optionally signing with
 * multiple secrets for zero-downtime key rotation.
 */
export class Signer {
  private readonly secrets: string[];

  constructor(secrets: string | string[]) {
    this.secrets = Array.isArray(secrets) ? secrets : [secrets];
  }

  /**
   * Computes the `webhook-id`, `webhook-timestamp`, and `webhook-signature` headers.
   * @param id - The webhook delivery id
   * @param payload - The exact raw JSON payload string
   * @param timestamp - The delivery timestamp (default: now)
   * @returns A record of the three `webhook-*` headers
   */
  headers(id: string, payload: string, timestamp: Date = new Date()): Record<string, string> {
    const timestampSeconds = Math.floor(timestamp.getTime() / 1000);
    const signatures = this.secrets.map((secret) => sign(secret, id, timestamp, payload));

    return {
      'webhook-id': id,
      'webhook-timestamp': String(timestampSeconds),
      'webhook-signature': signatures.join(' '),
    };
  }
}

/**
 * Structural shape matching anyhook-core's injected `WebhookSigner` port.
 */
interface WebhookSignerPort {
  sign(secrets: string[], id: string, payload: string, timestampMs: number): Record<string, string>;
}

/**
 * Creates a signer matching anyhook-core's `WebhookSigner` port shape.
 * anyhook-core injects this dependency; core itself does NOT import this package.
 * @returns A signer object with a `sign` method
 */
export function createSigner(): WebhookSignerPort {
  return {
    sign(secrets: string[], id: string, payload: string, timestampMs: number): Record<string, string> {
      return new Signer(secrets).headers(id, payload, new Date(timestampMs));
    },
  };
}
